optics-core/src/Optics/Traversal.hs

```diff
@@ -470,6 +478,59 @@ adjoin4 o1 o2 o3 o4 = combined % traversed
       []       ->            pure a
 {-# INLINE [1] adjoin4 #-}
 
+----------------------------------------
+
+-- | Traverse immediate occurrences of the type @a@ within the type @s@ using
+-- its 'Generic' instance.
+--
+-- >>> toListOf gplate ('h', ((), 'e', Just 'l'), "lo") :: String
+-- "hello"
+--
+-- /Note:/ types without the 'Generic' instance are ignored during the
+-- traversal.
+gplate :: (Generic s, GPlated (Rep s) a) => Traversal' s a
+gplate = traversalVL $ \f s -> to <$> gplated f (from s)
+{-# INLINE gplate #-}
+
+class GPlated g a where
+  gplated :: TraversalVL' (g p) a
+
+instance GPlated f a => GPlated (M1 i c f) a where
+  gplated f (M1 x) = M1 <$> gplated f x
+
+instance (GPlated f a, GPlated g a) => GPlated (f :+: g) a where
+  gplated f (L1 x) = L1 <$> gplated f x
+  gplated f (R1 x) = R1 <$> gplated f x
+
+instance (GPlated f a, GPlated g a) => GPlated (f :*: g) a where
+  gplated f (x :*: y) = (:*:) <$> gplated f x <*> gplated f y
+
+-- | Matching type.
+instance {-# OVERLAPPING #-} GPlated (K1 i a) a where
+  gplated f (K1 a) = K1 <$> f a
+
+-- | Recurse into the inner type if it has a 'Generic' instance.
+instance GPlateInner (HasRep (Rep b)) b a => GPlated (K1 i b) a where
+  gplated f (K1 b) = K1 <$> gplateInner @(HasRep (Rep b)) f b
+
+instance GPlated U1 a where
+  gplated _ = pure
+
+instance GPlated V1 a where
+  gplated _ = \case {}
+
+instance GPlated (URec b) a where
+  gplated _ = pure
+
+class GPlateInner (repDefined :: RepDefined) b a where
+  gplateInner :: TraversalVL' b a
+
+instance (Generic b, GPlated (Rep b) a) => GPlateInner 'RepDefined b a where
+  gplateInner f = fmap to . gplated f . from
+
+instance {-# INCOHERENT #-} GPlateInner repNotDefined b a where
+  gplateInner _ = pure
```

Comment:
This makes me a bit nervous. If I understand correctly, `gplate` in `lens` will not traverse any substructures nested inside other types, whereas the definition here will do so iff they have a `Generic` instance that is apparent at the call site. What happens if you have a polymorphic type that later gets instantiated?

Reply:
`GPlateInner` is hidden and it will not be resolved before `GPlated` is resolved, whereas `GPlated` will not resolve until the type is known, so it looks fine to me.

Reply:
Example:

```haskell
λ> let f = toListOf gplate :: (a, b) -> [a]

<interactive>:16:18: error:
    • Overlapping instances for GPlated
                                  25 (GHC.Generics.K1 GHC.Generics.R b1) a1
        arising from a use of ‘gplate’
      Matching instances:
        two instances involving out-of-scope types
          instance Optics.Traversal.GPlateInner
                     (Optics.Internal.Optic.TypeLevel.HasRep (GHC.Generics.Rep b))
                     (1 GHC.TypeNats.<=? n)
                     (n GHC.TypeNats.- 1)
                     b
                     a =>
                   GPlated n (GHC.Generics.K1 i b) a
            -- Defined at src/Optics/Traversal.hs:535:3
          instance [overlapping] GPlated n (GHC.Generics.K1 i a) a
            -- Defined at src/Optics/Traversal.hs:530:30
      (The choice depends on the instantiation of ‘b1, a1’
       To pick the first instance above, use IncoherentInstances
       when compiling the other instance declarations)
    • In the first argument of ‘toListOf’, namely ‘gplate’
      In the expression: toListOf gplate :: (a, b) -> [a]
      In an equation for ‘f’: f = toListOf gplate :: (a, b) -> [a]
```

I guess your concern was that if this was permitted and then `f` was used e.g. with `b ~ a`, right?

Reply:
Right, that's the kind of thing I was worried about. And the point is that we can't simplify `GPlated n (K1 i b) a` until we know that `a` is apart from `b`, which means we must have a concrete type for `b`, at which point we know whether it has a `Generic` instance... except if someone defines an orphan `Generic` instance later?

I haven't tested, but I think in the orphan case, one can have the "same" `gplate` call return different results because one has the `Generic` instance in scope and the other does not. Is that a deserved punishment for creating orphans?

Reply:
Yeah, if there is no `Generic` instance for the type in question in scope when `GPlated` is resolved, it won't traverse the type.

IMO documenting it should suffice, i.e. the note in `depth` needs to be refined.

Reply:
Yes, I'm okay with just documenting this.